### PR TITLE
Feature/42184/date picker banner

### PR DESCRIPTION
**Pull request by oliverguenther**
**Acceptance criteria**

- [x]   The banner can be in two different colours (or not existent at all)
- [x]   There can be two types of banners:
    - [x]   An **information banner** has a blue background
    - [x]  A **warning banner**  has an orange background
- [x]   There are 4 identified cases in which this banner is displayed:
    - [x]   **Warning:** Manual scheduling is enabled so all existing work package relations are ignored. (This is only visible when the work package in question has relations; a manually scheduled work package with no relations will not show this banner).
    - [x]   **Warning:** Changing the dates of this work package will change the dates of related work packages
    - [x]  **Information:** Available start and finish dates are limited by relations.
    - [x]  **Information:** The dates are automatically derived from relations (and date input is disabled).
- [x]   Structure (may already be done in <mention class="mention" data-id="41814" data-type="work_package" data-text="#41814">#41814</mention> )
    - [x]   an icon
    - [x]   two-line message, the first in bold
    - [x]   a &quot;Show relations&quot; button that, when clicked, opens up a GANTT preview in a new window
        - [x]   This is a filtered work package list (table view + GANTT view) showing only work packages related to the &quot;main&quot; work package (the one for which the GANTT preview was requested).
    - [x]   The mobile date picker displays a simplified version of the banner:
        - [x]   Without the left-side icon
        - [x]   Without the second line of text (only the first line is displayed)
        - [x]  With the usual &quot;Show relations&quot; button.

https://community.openproject.org/wp/42184

### CODE REVIEW

File: config/locales/js-en.yml [hunk 1]
```diff
@@ -924,6 +924,13 @@ en:
       comment_send_failed: "An error has occurred. Could not submit the comment."
       comment_updated: "The comment was successfully updated."
       confirm_edit_cancel: "Are you sure you want to cancel editing the work package?"
+      datepicker_modal:
+        automatically_scheduled_parent: "Automatically scheduled. Dates are derived from relations."
+        manually_scheduled: "Manual scheduling enabled, all relations ignored."
+        start_date_limited_by_relations: "Available start date is limited by relations."
+        changing_dates_affects_follow_relations: "Changing these dates will affect dates of related work packages."
```
Comment (HDinger) [inline config/locales/js-en.yml]: The text differs from the specification.

File: frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    grid-template-columns: auto max-content
+
+  &_warning
+    background: $spot-color-feedback-warning-light
+
+  &_info
+    background: $spot-color-feedback-info-light
+
+  &--text
+    margin: 0
```
Comment (HDinger) [inline frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass]: The `.op-modal-banner--text` as well as the `.op-modal-banner--title` are too big compared to the visuals. Taken from figma, it should be a `caption` (in bold and regular).

You can simply use the `spot-caption` mixin for that. You can also pass `bold` as an argument for the font-weight. The special rules for these two classes should then not be necessary any more.

File: frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    line-height: 1
+
+  &--icon
+    font-size: 20px
```
Comment (HDinger) [inline frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass]: That is unnecessary as the specification says 16px (which is the default)

File: frontend/src/app/shared/components/datepicker/datepicker.modal.sass [hunk 4]
```diff
@@ -8,6 +8,9 @@
   max-width: 730px
   min-height: 200px
   box-shadow: $spot-shadow-light-mid
+  // Apply a bottom margin before the banners are loaded
+  // to avoid overlapping
+  margin-bottom: 60px
```
Comment (HDinger) [inline frontend/src/app/shared/components/datepicker/datepicker.modal.sass:13]: I don't really get the point of this.. If I remove this line, I don't see any difference. Is it a timing issue?

**Review by HDinger — CHANGES_REQUESTED** [review]
I have some remarks:

- [x] When the scheduling is disabled, the datepicker is not shown (but it should be, in the disabled mode). The "disabled" mode of the datepicker with all belonging styles is automatically set when [this methods](https://github.com/opf/openproject/blob/bbc9a351406a04064a7acb3611f08a0e8eb804cf/frontend/src/app/shared/components/datepicker/datepicker.modal.ts#L221) returns `false`.
- [x] On mobile, the information banner looks off (the icon is still shown, and the text not wrapped). However, when I click on "manual scheduling" the banner is reloaded, and then looks correctly.

Bildschirmfoto 2022-06-07 um 12 47 22

- [x] I am not quite sure about this one, but when I open a child work package, which is the last in the hierarchy, shouldn't then be the warning shown "Changing the dates of this work package will change the dates of related work packages"
Bildschirmfoto 2022-06-07 um 12 49 34

File: frontend/src/app/shared/components/datepicker/datepicker.modal.sass (hunk 4 above)
Comment (oliverguenther) [reply frontend/src/app/shared/components/datepicker/datepicker.modal.sass:13]: The way this occurs is as follows:

1. You open the datepicker for a WP that has a precedes/follows relation
2. The datepicker modal opens briefly without banner
3. The banner information gets loaded, showing the banner

This pushes the modal to the bottom, which might result in it overlapping the window border and being cut off IF it was opened near the bottom.

File: frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass (hunk 3 above)
Comment (oliverguenther) [reply frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.sass]: Maybe I'm not clear how this works, as I've seen 16px being used in Figma, but this results in wildly different appearances in app.

**Review by HDinger — APPROVED** [review]

File: frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.html [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<div
+  class="op-modal-banner"
+  ngClass="op-modal-banner_{{ type }}"
+  attr.data-qa-selector="op-modal-banner-{{ type }}"
+>
+  <span
+    *ngIf="type === 'info'"
+    class="op-modal-banner--icon spot-icon spot-icon_info1 hidden-for-mobile"
```
Comment (b12f) [inline frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.html]: I don't see the `--icon` element being selected in CSS. Maybe we can just remove it?

**Review by b12f — APPROVED** [review]
LGTM

File: frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.html (hunk 5 above)
Comment (HDinger) [reply frontend/src/app/shared/components/modal/modal-banner/modal-banner.component.html]: Done 👍

**Comment by HDinger** [comment]
I addressed b12f's comment. This is good to merge once the tests have run through 👍

**Pull request merged**